### PR TITLE
[improvement](memory) Storage page cache use LRU-K cache, K=2

### PR DESCRIPTION
**Pull request by xinyiZzz**
### What problem does this PR solve?

Storage page cache uses plain LRU Cache, occasional batch operations can cause "cache pollution" in plain LRU Cache. This will cause hotspot data to be squeezed out of the cache by non-hotspot data, reduce cache hit rate.

In extreme cases, if the number of pages inserted each time is greater than the cache capacity, the cache hit rate will be 0.

Introducing LRU-K Cache avoids "cache pollution" in most cases.

Last submitted #28794
Similar PR: https://github.com/apache/doris/pull/23546

Performance test on branch-2.1:
1-FE, 3-BE (32 Core, 128GB Memory) Doris cluster
`ssb_sf_1000` all query cost: 18.85s -> 14.44s, 23% improvement
`ssb_flat_1000` all query cost: 12.21s -> 8.54s, 30% improvement, cache hit rate 1% -> 40%

### Check List (For Author)

- Test 
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason 

- Behavior changed:
    - [ ] No.
    - [ ] Yes. 

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. 

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by Thearas** [comment]
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

File: be/src/olap/lru_cache.cpp [hunk 1]
```diff
@@ -419,13 +474,18 @@ Cache::Handle* LRUCache::insert(const CacheKey& key, uint32_t hash, void* value,
...
+    e->refs = 1; // only one for the returned handle.
     e->next = e->prev = nullptr;
-    e->in_cache = true;
+    e->in_cache = false;
     e->priority = priority;
     e->type = _type;
     memcpy(e->key_data, key.data(), key.size());
     e->last_visit_time = UnixMillis();
+
+    if (_is_lru_k && _lru_k_insert_visits_list(e->total_size, hash)) {
```
Comment (yiguolei) [inline be/src/olap/lru_cache.cpp:485]: 为啥这里是判断hash 值，而不是判断key？

File: be/src/olap/lru_cache.cpp [hunk 2]
```diff
@@ -476,14 +545,15 @@ void LRUCache::erase(const CacheKey& key, uint32_t hash) {
         e = _table.remove(key, hash);
         if (e != nullptr) {
             last_ref = _unref(e);
-            if (last_ref) {
```
Comment (yiguolei) [inline be/src/olap/lru_cache.cpp:479]: 原来的代码里，如果478 行，last ref == false，什么时候才执行480 行的代码呢？

Comment (xinyiZzz) [reply be/src/olap/lru_cache.cpp:479]: 这里，等最后一个持有entry的 handle release 的时候
image

File: be/src/olap/lru_cache.cpp (hunk 1 above)
Comment (xinyiZzz) [reply be/src/olap/lru_cache.cpp:485]: 因为 key 可能很长，key的长度是没有限制的，所以是先比较 hash 值，再比较 key
image

File: be/src/olap/lru_cache.cpp [hunk 3]
```diff
@@ -289,21 +289,35 @@ void LRUCache::_lru_append(LRUHandle* list, LRUHandle* e) {
...
+            e->last_visit_time = UnixMillis();
+        } else {
+            ++_miss_count;
+        }
+    }
+    // If key not exist in cache, and is lru k cache, and key in visits list,
+    // then move the key to beginning of the visits list.
+    // key in visits list indicates that the key has been inserted once after the cache is full.
+    if (e == nullptr && _is_lru_k) {
+        std::lock_guard l(_visits_lru_cache_mutex);
```
Comment (yiguolei) [inline be/src/olap/lru_cache.cpp:315]: 这个锁的意义感觉不大，直接使用跟LRUCache 一个_mutex 也没问题

Comment (xinyiZzz) [reply be/src/olap/lru_cache.cpp:315]: 最好2个锁，现在 lru cache 锁的开销在某些场景下已经很大了，比如 prune 淘汰的时候，
_visits_lru_cache_list 是个完全独立的数据结构，这样第一次insert的时候可以不用拿 cache _mutex
image

**Comment by xinyiZzz** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]
LGTM

**Review by wangbo — APPROVED** [review]
LGTM

**Pull request merged**